### PR TITLE
Chisel decoder

### PR DESCRIPTION
**Pull request by sequencer**
This PR use new chisel API to replace original RocketChip Decoder.
Chisel Decoder depends on the explicit width, so I adds explicit width for most usages. This will be a breaking change to RC.
**Related issue**: 

**Type of change**: new feature

**Impact**: no functional change

**Development Phase**: implementation

**Release Notes**
Switch to Chisel Decoder

### CODE REVIEW

**Comment by sequencer** [comment]
@jackkoenig Ask to merge this.

**Comment by jackkoenig** [comment]
I bumped to 3.5-SNAPSHOT to show that this is still failing (at least for me locally)

**Comment by sequencer** [comment]
This is really wired:
```
  BitPat(0000001??????????000?????0110011) -> BitPat(1000001101001???100?????00000110000000)
  BitPat(0000001??????????001?????0110011) -> BitPat(1000001101001???110?????00000110000000)
  BitPat(0000001??????????011?????0110011) -> BitPat(1000001101001???1110?????00000110000000)
  BitPat(0000001??????????010?????0110011) -> BitPat(1000001101001???1100?????00000110000000)
  BitPat(0000001??????????100?????0110011) -> BitPat(1000001101001???11000?????00000110000000)
  BitPat(0000001??????????101?????0110011) -> BitPat(1000001101001???11010?????00000110000000)
  BitPat(0000001??????????110?????0110011) -> BitPat(1000001101001???11100?????00000110000000)
  BitPat(0000001??????????111?????0110011) -> BitPat(1000001101001???11110?????00000110000000)
  BitPat(00000????????????010?????0101111) -> BitPat(1000001100001???101100000000010000010)
  BitPat(00100????????????010?????0101111) -> BitPat(1000001100001???101100100000010000010)
  BitPat(00001????????????010?????0101111) -> BitPat(1000001100001???10110000000010000010)
  BitPat(01100????????????010?????0101111) -> BitPat(1000001100001???101101100000010000010)
  BitPat(01000????????????010?????0101111) -> BitPat(1000001100001???101101000000010000010)
  BitPat(10000????????????010?????0101111) -> BitPat(1000001100001???101110000000010000010)
  BitPat(11000????????????010?????0101111) -> BitPat(1000001100001???101111000000010000010)
  BitPat(10100????????????010?????0101111) -> BitPat(1000001100001???101110100000010000010)
  BitPat(11100????????????010?????0101111) -> BitPat(1000001100001???101111100000010000010)
  BitPat(00010??00000?????010?????0101111) -> BitPat(1000001100001???10111000000010000010)
  BitPat(00011????????????010?????0101111) -> BitPat(1000001100001???10111100000010000010)
  BitPat(0010000??????????000?????1010011) -> BitPat(110000000????????????0?????11010000000000)
  BitPat(0010000??????????010?????1010011) -> BitPat(110000000????????????0?????11010000000000)
  BitPat(0010000??????????001?????1010011) -> BitPat(110000000????????????0?????11010000000000)
  BitPat(0010100??????????000?????1010011) -> BitPat(110000000????????????0?????11010000000000)
  BitPat(0010100??????????001?????1010011) -> BitPat(110000000????????????0?????11010000000000)
  BitPat(0000000??????????????????1010011) -> BitPat(110000000????????????0?????11010000000000)
  BitPat(0000100??????????????????1010011) -> BitPat(110000000????????????0?????11010000000000)
  BitPat(0001000??????????????????1010011) -> BitPat(110000000????????????0?????11010000000000)
  BitPat(?????00??????????????????1000011) -> BitPat(110000000????????????0?????11110000000000)
  BitPat(?????00??????????????????1000111) -> BitPat(110000000????????????0?????11110000000000)
  BitPat(?????00??????????????????1001111) -> BitPat(110000000????????????0?????11110000000000)
  BitPat(?????00??????????????????1001011) -> BitPat(110000000????????????0?????11110000000000)
  BitPat(111000000000?????001?????1010011) -> BitPat(110000000????????????0?????10000010000000)
  BitPat(111000000000?????000?????1010011) -> BitPat(110000000????????????0?????10000010000000)
  BitPat(110000000000?????????????1010011) -> BitPat(110000000????????????0?????10000010000000)
  BitPat(110000000001?????????????1010011) -> BitPat(110000000????????????0?????10000010000000)
  BitPat(1010000??????????010?????1010011) -> BitPat(110000000????????????0?????11000010000000)
  BitPat(1010000??????????001?????1010011) -> BitPat(110000000????????????0?????11000010000000)
  BitPat(1010000??????????000?????1010011) -> BitPat(110000000????????????0?????11000010000000)
  BitPat(111100000000?????000?????1010011) -> BitPat(110000010??01????????0?????00010000000000)
  BitPat(110100000000?????????????1010011) -> BitPat(110000010??01????????0?????00010000000000)
  BitPat(110100000001?????????????1010011) -> BitPat(110000010??01????????0?????00010000000000)
  BitPat(?????????????????010?????0000111) -> BitPat(1100000101101100101000010000000000)
  BitPat(?????????????????010?????0100111) -> BitPat(1100000101101000101101000000000000)
  BitPat(0001100??????????????????1010011) -> BitPat(110000000????????????0?????11010000000000)
  BitPat(010110000000?????????????1010011) -> BitPat(110000000????????????0?????11010000000000)
  BitPat(0000000??????????001?????0010011) -> BitPat(1000000101101100110?????00000010000000)
  BitPat(0000000??????????101?????0010011) -> BitPat(100000010110110011010?????00000010000000)
  BitPat(0100000??????????101?????0010011) -> BitPat(1000000101101100110110?????00000010000000)
  BitPat(0001001??????????000000001110011) -> BitPat(1000001100001???1011010000000001000000)
  BitPat(00010000001000000000000001110011) -> BitPat(1000000?0????????????0?????00000001000000)
  BitPat(01111011001000000000000001110011) -> BitPat(1000000?0????????????0?????00000001000000)
  BitPat(?????????????????001?????0001111) -> BitPat(100000000????????????0?????00000000001100)
  BitPat(?????????????????001?????1100011) -> BitPat(1001001101001001?110?????00000000000000)
  BitPat(?????????????????000?????1100011) -> BitPat(1001001101001001?100?????00000000000000)
  BitPat(?????????????????100?????1100011) -> BitPat(1001001101001001?11000?????00000000000000)
  BitPat(?????????????????110?????1100011) -> BitPat(1001001101001001?11100?????00000000000000)
  BitPat(?????????????????101?????1100011) -> BitPat(1001001101001001?11010?????00000000000000)
  BitPat(?????????????????111?????1100011) -> BitPat(1001001101001001?11110?????00000000000000)
  BitPat(?????????????????????????1101111) -> BitPat(1000100000110011100?????00000010000000)
  BitPat(?????????????????000?????1100111) -> BitPat(1000010101101100100?????00000010000000)
  BitPat(?????????????????????????0010111) -> BitPat(1000000001110010100?????00000010000000)
  BitPat(?????????????????000?????0000011) -> BitPat(1000000101101100101000000010000000)
  BitPat(?????????????????001?????0000011) -> BitPat(1000000101101100101000000010000000)
  BitPat(?????????????????010?????0000011) -> BitPat(1000000101101100101000000010000000)
  BitPat(?????????????????100?????0000011) -> BitPat(1000000101101100101000000010000000)
  BitPat(?????????????????101?????0000011) -> BitPat(1000000101101100101000000010000000)
  BitPat(?????????????????000?????0100011) -> BitPat(1000001101101000101100000000000000)
  BitPat(?????????????????001?????0100011) -> BitPat(1000001101101000101100000000000000)
  BitPat(?????????????????010?????0100011) -> BitPat(1000001101101000101100000000000000)
  BitPat(?????????????????????????0110111) -> BitPat(1000000001100010100?????00000010000000)
  BitPat(?????????????????000?????0010011) -> BitPat(1000000101101100100?????00000010000000)
  BitPat(?????????????????010?????0010011) -> BitPat(1000000101101100111000?????00000010000000)
  BitPat(?????????????????011?????0010011) -> BitPat(1000000101101100111100?????00000010000000)
  BitPat(?????????????????111?????0010011) -> BitPat(100000010110110011110?????00000010000000)
  BitPat(?????????????????110?????0010011) -> BitPat(100000010110110011100?????00000010000000)
  BitPat(?????????????????100?????0010011) -> BitPat(100000010110110011000?????00000010000000)
  BitPat(0000000??????????000?????0110011) -> BitPat(1000001101001???100?????00000010000000)
  BitPat(0100000??????????000?????0110011) -> BitPat(1000001101001???110100?????00000010000000)
  BitPat(0000000??????????010?????0110011) -> BitPat(1000001101001???111000?????00000010000000)
  BitPat(0000000??????????011?????0110011) -> BitPat(1000001101001???111100?????00000010000000)
  BitPat(0000000??????????111?????0110011) -> BitPat(1000001101001???11110?????00000010000000)
  BitPat(0000000??????????110?????0110011) -> BitPat(1000001101001???11100?????00000010000000)
  BitPat(0000000??????????100?????0110011) -> BitPat(1000001101001???11000?????00000010000000)
  BitPat(0000000??????????001?????0110011) -> BitPat(1000001101001???110?????00000010000000)
  BitPat(0000000??????????101?????0110011) -> BitPat(1000001101001???11010?????00000010000000)
  BitPat(0100000??????????101?????0110011) -> BitPat(1000001101001???110110?????00000010000000)
  BitPat(?????????????????000?????0001111) -> BitPat(100000000????????????0?????00000000000100)
  BitPat(00000000000000000000000001110011) -> BitPat(1000000?0????????????0?????00000001000000)
  BitPat(00000000000100000000000001110011) -> BitPat(1000000?0????????????0?????00000001000000)
  BitPat(00110000001000000000000001110011) -> BitPat(1000000?0????????????0?????00000001000000)
  BitPat(00010000010100000000000001110011) -> BitPat(1000000?0????????????0?????00000001000000)
  BitPat(00110000010100000000000001110011) -> BitPat(1000000?0????????????0?????00000001000000)
  BitPat(?????????????????001?????1110011) -> BitPat(1000000100001???100?????00000011010000)
  BitPat(?????????????????010?????1110011) -> BitPat(1000000100001???100?????00000011100000)
  BitPat(?????????????????011?????1110011) -> BitPat(1000000100001???100?????00000011110000)
  BitPat(?????????????????101?????1110011) -> BitPat(1000000001100101100?????00000011010000)
  BitPat(?????????????????110?????1110011) -> BitPat(1000000001100101100?????00000011100000)
  BitPat(?????????????????111?????1110011) -> BitPat(1000000001100101100?????00000011110000)
```
The Rocket Decoder do have different output width...
However that should be impossible...

File: src/main/scala/rocket/ALU.scala [hunk 1]
```diff
@@ -10,21 +10,21 @@ import freechips.rocketchip.tile.CoreModule
...
+  def FN_XOR  = 4.U(4.W)
+  def FN_SR   = 5.U(4.W)
+  def FN_OR   = 6.U(4.W)
+  def FN_AND  = 7.U(4.W)
+  def FN_SUB  = 10.U(4.W)
+  def FN_SRA  = 11.U(4.W)
+  def FN_SLT  = 12.U(4.W)
+  def FN_SGE  = 13.U(4.W)
+  def FN_SLTU = 14.U(4.W)
+  def FN_SGEU = 15.U(4.W)
```
Comment (sequencer) [inline src/main/scala/rocket/ALU.scala:27]: Why this decoding result without explicit width works before? Confusing...

Comment (sequencer) [reply src/main/scala/rocket/ALU.scala:27]: infer width with default? I don't know...

**Comment by sequencer** [comment]
ae002f5f9d11f1bb865f766cd14d2b8bc28b111d fixed a legacy bug from original QMC decoder in RocketChip. Thanks debugging by @ZenithalHourlyRate

**Comment by sequencer** [comment]
@jackkoenig Ask for merge.

**Comment by sequencer** [comment]
It works with 3.5.1 now. After review, I'll rebase it.

File: src/main/scala/rocket/ALU.scala (hunk 1 above)
Comment (jackkoenig) [reply src/main/scala/rocket/ALU.scala:27]: As of https://github.com/chipsalliance/chisel3/pull/2387 (which has been backported to 3.5), these widths should no longer be necessary right? Can you change it back to minimize diff?

File: src/main/scala/rocket/Consts.scala [hunk 2]
```diff
@@ -2,80 +2,81 @@
...
+  def M_XA_MAXU = "b01111".U(5.W)
+  def M_FLUSH   = "b10000".U(5.W) // write back dirty data and cede R/W permissions
+  def M_PWR     = "b10001".U(5.W) // partial (masked) store
+  def M_PRODUCE = "b10010".U(5.W) // write back dirty data and cede W permissions
+  def M_CLEAN   = "b10011".U(5.W) // write back dirty data and retain R/W permissions
+  def M_SFENCE  = "b10100".U(5.W) // SFENCE.VMA
+  def M_HFENCEV = "b10101".U(5.W) // HFENCE.VVMA
+  def M_HFENCEG = "b10110".U(5.W) // HFENCE.GVMA
+  def M_WOK     = "b10111".U(5.W) // check write permissions but don't perform a write
+  def M_HLVX    = "b10000".U(5.W) // HLVX instruction
```
Comment (jackkoenig) [inline src/main/scala/rocket/Consts.scala:80]: All widths in this file should similarly no longer be necessary right? Can we undo the change just to minimize the diff?

File: src/main/scala/tilelink/Bundles.scala [hunk 3]
```diff
@@ -15,26 +15,26 @@ abstract class TLBundleBase(params: TLBundleParameters) extends GenericParameter
...
+  def AccessAckData  = 1.U(3.W) //               .    .
+  def HintAck        = 2.U(3.W) //               .    .
+  def ProbeAck       = 4.U(3.W) //               .
+  def ProbeAckData   = 5.U(3.W) //               .
+  def Release        = 6.U(3.W) //               .              => ReleaseAck
+  def ReleaseData    = 7.U(3.W) //               .              => ReleaseAck
+  def Grant          = 4.U(3.W) //                    .         => GrantAck
+  def GrantData      = 5.U(3.W) //                    .         => GrantAck
+  def ReleaseAck     = 6.U(3.W) //                    .
+  def GrantAck       = 0.U(3.W) //                         .
```
Comment (jackkoenig) [inline src/main/scala/tilelink/Bundles.scala:37]: Same width complaint here.

File: src/main/scala/rocket/Consts.scala (hunk 2 above)
Comment (sequencer) [reply src/main/scala/rocket/Consts.scala:80]: Yes, but it seems that chipsalliance/chisel3#2387 didn't catch the 3.5.1...

**Review by jackkoenig — DISMISSED** [review]
Speaking with rocket-chip folks, it would be better to not need the explicit widths, let's get 3.5.2 released and bump to that which will make this work better.

**Comment by jackkoenig** [comment]
@sequencer I've pushed some substantial changes that remove the need for any explicit widths. Now `DecodeLogic` is safe to differing widths from the `BitPats` (and will pad them so long as they have no don't cares). Since I've now contributed substantially to this, I'm going to dismiss my review and request review from @aswaterman

File: build.sbt [hunk 4]
```diff
@@ -5,7 +5,7 @@ import scala.sys.process._
 
 enablePlugins(PackPlugin)
 
-val chiselVersion = "3.5-SNAPSHOT"
+val chiselVersion = "3.5.0"
```
Comment (michael-etzkorn) [inline build.sbt:8]: FWIW, main is currently using 3.5.2 and this would revert that.

Comment (michael-etzkorn) [reply build.sbt:8]: Oh, I guess with the way the commit history works, this just ends up as a no-change so never mind me here.

Comment (jackkoenig) [reply build.sbt:8]: I think you're commenting on a commit, I was basically reverting a prior commit so that the overall PR makes no changes to `build.sbt`.

**Comment by michael-etzkorn** [comment]
Ignore my review / comments / deleted comments, the `build.sbt` commit changes aren't relevant to the PR

**Comment by jackkoenig** [comment]
For posterity who may hit the requirement about a `BitPat` having don't cares and needing padding, see https://sifive.zoom.us/j/96196969440?pwd=WXlJZHpjWkdTUkNBS0FFbWRhSUlHZz09 for an example of the general approach.

**Comment by jackkoenig** [comment]
For posterity who may hit the requirement about a BitPat having don't cares and needing padding, see https://github.com/chipsalliance/rocket-chip/pull/2949 for an example of the general approach.

**Review by aswaterman — DISMISSED** [review]
Ship it!

**Comment by sequencer** [comment]
opps, I doubt this may blames to CSR addresses decoding.

**Comment by jackkoenig** [comment]
> opps, I doubt this may blames to CSR addresses decoding.

@sequencer I've restored one of your changes to the CSR in https://github.com/chipsalliance/rocket-chip/pull/2836/commits/f2cbe128f819c6c86314d10947a745d369166637, do you think this is the issue?

**Comment by sequencer** [comment]
Yes, I have been thinking, I thought I used to be a bug, but I can't understand, why it worked before...

**Review by aswaterman — APPROVED** [review]
This is fine with me.  I was relying on QM doing the optimal thing (there should be no dependence on input bits 19..0 in an optimized circuit).
